src/System.Linq/tests/ToArrayTests.cs

```diff
@@ -199,5 +199,95 @@ public void ToArray_ListWhereSelect(int[] sourceIntegers, string[] convertedStri
             Assert.Equal(Array.Empty<string>(), sourceList.Select(i => i.ToString()).Where(s => s == null).ToArray());
         }
 
+        [Fact]
+        public void SameResultsRepeatCallsFromWhereOnIntQuery()
+        {
+            var q = from x in new[] { 9999, 0, 888, -1, 66, -777, 1, 2, -12345 }
+                    where x > Int32.MinValue
+                    select x;
+
+            Assert.Equal(q.ToArray(), q.ToArray());
+        }
+        
+        [Fact]
+        public void SameResultsRepeatCallsFromWhereOnStringQuery()
+        {
+            var q = from x in new[] { "!@#$%^", "C", "AAA", "", "Calling Twice", "SoS", String.Empty }
+                        where !String.IsNullOrEmpty(x)
+                        select x;
+
+            Assert.Equal(q.ToArray(), q.ToArray());
+        }
+
+        [Fact]
+        public void SourceIsEmptyICollectionT()
+        {
+            int[] source = { };
+            int[] expected = { };
+
+            ICollection<int> collection = source as ICollection<int>;
+
+            Assert.Equal(expected, source.ToArray());
+            Assert.Equal(expected, collection.ToArray());
+        }
+
+        [Fact]
+        public void SourceIsICollectionTWithFewElements()
+        {
+            int?[] source = { -5, null, 0, 10, 3, -1, null, 4, 9 };
+            int?[] expected = { -5, null, 0, 10, 3, -1, null, 4, 9 };
+
+            ICollection<int?> collection = source as ICollection<int?>;
```

Comment:
This test is a bit odd.  The only overload of ToArray is strongly-typed to take an ```IEnumerable<T>```, so this doesn't impact what function is called, and the implementation does its own internal casting to check for special interfaces, including ```ICollection<T>```.  In case ToArray ever added specialized code paths for ```T[]``` rather than just for ```ICollection<T>```, I'd think it'd be better if this was some ```ICollection<T>``` that's not a ```T[]```, like a ```List<T>```, but we can leave it this way for now to stay true to the original test, and update it as needed I guess.